gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java

```diff
@@ -764,6 +776,34 @@ private void renameDirectoryInternal(FileInfo srcInfo, URI dst) throws IOExcepti
       }
     }
 
+    StorageResourceId srcResourceId =
+        StorageResourceId.fromUriPath(src, /* allowEmptyObjectName= */ true);
+    StorageResourceId dstResourceId =
+        StorageResourceId.fromUriPath(
+            dst, /* allowEmptyObjectName= */ true, /* generationId= */ 0L);
+    if (this.options.getCloudStorageOptions().isMoveOperationEnabled()
+        && srcResourceId.getBucketName().equals(dstResourceId.getBucketName())) {
+
+      // First, move all items except marker items
+      moveInternal(srcToDstItemNames);
+      // Finally, move marker items (if any) to mark rename operation success
+      moveInternal(srcToDstMarkerItemNames);
+
+      List<FileInfo> bucketToDelete = new ArrayList<>();
+      List<FileInfo> objectToDelete = new ArrayList<>();
+
+      if (srcInfo.getItemInfo().isBucket()) {
+        bucketToDelete.add(srcInfo);
+        deleteBucket(bucketToDelete);
+      } else {
+        // If src is a directory then srcItemInfos does not contain its own name,
+        // we delete item separately in the list.
+        objectToDelete.add(srcInfo);
+        deleteObjects(objectToDelete);
```

Comment:
It seems we are deleting the srcInfo, if its an object we call deleteObject if it's a bucket we call deleteBucket. This whole logic of creating a list and then adding item to can be simplified.